app/controllers/api/observation_sessions_controller.rb

```diff
@@ -2,16 +2,24 @@ module Api
   class ObservationSessionsController < BaseController
     def create
       ActiveRecord::Base.transaction do
-        create_observation_sessions
-        # TODO: upsert?
-        observations = observations_params.map do |observation|
+        
+      session_ids_to_instances = {} 
+      processed_observations = []   
+
+      observations_params.each do | observation |
+          session = session_ids_to_instances.fetch(observation['observation_session_id']) do | session_id |
+            session_ids_to_instances[session_id] = ObservationSession.find_or_create_by(id: session_id, user_id: current_user.id)
+          end
+          
           Observation.find_or_create_by(id: observation['id']) do |record|
             record.details = observation
-            record.observation_session_id = observation['observation_session_id']
+            record.observation_session = session
+            record.save
```

Comment:
Hmm, that's a good point.  I think since the data is immutable from the device that this is OKAY.  It's certainly something that we might want to note.   Does leaving the save their trigger a double save?